### PR TITLE
Update to v4.0.0

### PR DESCRIPTION
**Pull request by viceroypenguin**

### CODE REVIEW

File: Directory.Packages.props [hunk 1]
```diff
@@ -1,42 +1,43 @@
...
-		<PackageVersion Include="System.Memory"                             Version="4.5.4"                                   />
+		<PackageVersion Include="System.Memory"                             Version="4.5.5"                                   />
 		<PackageVersion Include="System.Buffers"                            Version="4.5.1"                                   />
 		<PackageVersion Include="System.Configuration.ConfigurationManager" Version="6.0.0"                                   />
-		<PackageVersion Include="MySqlConnector"                            Version="2.0.0"                                   />
+		<PackageVersion Include="MySqlConnector"                            Version="2.1.10"                                  />
 		<PackageVersion Include="AdoNetCore.AseClient"                      Version="0.19.2"                                  />
 		<PackageVersion Include="IBM.Data.DB.Provider"                      Version="11.5.5010.4" GeneratePathProperty="true" />
 		<PackageVersion Include="System.Data.SQLite.Core"                   Version="1.0.115.5"                               />
```
Comment (MaceWindu) [inline Directory.Packages.props:18]: not in sync with nuspec

File: Directory.Packages.props [hunk 2]
```diff
@@ -1,42 +1,43 @@
...
 		<PackageVersion Include="System.Data.SqlClient"                     Version="4.8.3"                                   />
-		<PackageVersion Include="Oracle.ManagedDataAccess"                  Version="19.13.0"                                 />
-		<PackageVersion Include="Oracle.ManagedDataAccess.Core"             Version="3.21.61"                                  />
+		<PackageVersion Include="Oracle.ManagedDataAccess.Core"             Version="3.21.61"                                 />
 		<PackageVersion Include="System.Data.Odbc"                          Version="6.0.0"                                   />
 		<PackageVersion Include="System.Data.OleDb"                         Version="6.0.0"                                   />
 		<PackageVersion Include="IBM.Data.DB2.Core"                         Version="3.1.0.500"                               />
```
Comment (MaceWindu) [inline Directory.Packages.props:23]: not in sync with nuspec

File: Directory.Packages.props [hunk 3]
```diff
@@ -1,42 +1,43 @@
...
 	
 	<ItemGroup Condition=" '$(TargetFramework)' != 'net461' ">
-		<PackageVersion Include="Microsoft.CodeAnalysis.CSharp"   Version="4.0.1" />
-		<PackageVersion Include="Npgsql"                          Version="6.0.0" />
-		<PackageVersion Include="FirebirdSql.Data.FirebirdClient" Version="8.5.4" />
-		<PackageVersion Include="System.Text.Json"                Version="6.0.0" />
+		<PackageVersion Include="Microsoft.CodeAnalysis.CSharp"   Version="4.2.0"  />
+		<PackageVersion Include="Npgsql"                          Version="6.0.4"  />
+		<PackageVersion Include="FirebirdSql.Data.FirebirdClient" Version="9.0.1"  />
+		<PackageVersion Include="Oracle.ManagedDataAccess"        Version="21.6.1" />
```
Comment (MaceWindu) [inline Directory.Packages.props:40]: not in sync with nuspec

File: Source/SchemaAndCodeGenerator.cs [hunk 4]
```diff
@@ -465,13 +465,12 @@ ExplorerItem GetProcedures(string header, ExplorerIcon icon, List<ProcedureSchem
...
+						var sprocSqlName = _sqlBuilder!.BuildObjectName(
 							new StringBuilder(),
-							null,
-							null,
-							p.SchemaName == null ? null : _sqlBuilder.ConvertInline(p.SchemaName, ConvertType.NameToSchema),
-							_sqlBuilder.ConvertInline(p.ProcedureName, ConvertType.NameToQueryTable),
-							TableOptions.NotSet).ToString();
+							new SqlQuery.SqlObjectName(
+								Name: _sqlBuilder.ConvertInline(p.ProcedureName, ConvertType.NameToProcedure), 
+								Schema: p.SchemaName == null ? null : _sqlBuilder.ConvertInline(p.SchemaName, ConvertType.NameToSchema)),
```
Comment (MaceWindu) [inline Source/SchemaAndCodeGenerator.cs:472]: SqlObjectName contains raw name and convert handled by `BuildObjectName` method

File: Source/SchemaAndCodeGenerator.cs [hunk 5]
```diff
@@ -696,13 +695,12 @@ ExplorerItem GetTables(string header, ExplorerIcon icon, IEnumerable<TableSchema
...
+						var tableSqlName = _sqlBuilder!.BuildObjectName(
 							new StringBuilder(),
-							null,
-							null,
-							t.SchemaName == null ? null : _sqlBuilder.ConvertInline(t.SchemaName, ConvertType.NameToSchema),
-							_sqlBuilder.ConvertInline(t.TableName!,  ConvertType.NameToQueryTable),
-							TableOptions.NotSet).ToString();
+							new SqlQuery.SqlObjectName(
+								Name: _sqlBuilder.ConvertInline(t.TableName!, ConvertType.NameToQueryTable),
+								Schema: t.SchemaName == null ? null : _sqlBuilder.ConvertInline(t.SchemaName, ConvertType.NameToSchema)),
```
Comment (MaceWindu) [inline Source/SchemaAndCodeGenerator.cs:702]: same here, no manual convert required with BuildObjectName

**Review by MaceWindu — CHANGES_REQUESTED** [review]

File: Directory.Packages.props (hunk 1 above)
Comment (viceroypenguin) [reply Directory.Packages.props:18]: I blame you. ;)

File: Build/linq2db.LINQPad.nuspec [hunk 6]
```diff
@@ -11,26 +11,27 @@
...
-				<dependency id="System.Data.SQLite.Core"                   version="1.0.115.5" />
+				<dependency id="System.Data.SQLite.Core"                   version="1.0.116"   />
 				<dependency id="System.Data.SqlClient"                     version="4.8.3"     />
 				<dependency id="System.Data.Odbc"                          version="5.0.0"     />
 				<dependency id="System.Data.OleDb"                         version="5.0.0"     />
-				<dependency id="Npgsql"                                    version="6.0.0"     />
-				<dependency id="Oracle.ManagedDataAccess.Core"             version="3.21.4"    />
-				<dependency id="IBM.Data.DB2.Core"                         version="3.1.0.500" />
+				<dependency id="Npgsql"                                    version="6.0.4"     />
+				<dependency id="Oracle.ManagedDataAccess"                  version="21.6.1"    />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:32]: nuspec is for linqpad 6/7 targeting .net core. No need for netfx Oracle.ManagedDataAccess nuget

File: Build/linq2db.LINQPad.nuspec [hunk 7]
```diff
@@ -11,26 +11,26 @@
 		<readme>README.md</readme>
 		<dependencies>
 			<group targetFramework=".NETCoreApp3.1">
-				<dependency id="LINQPad.Reference"                         version="1.1.0"     />
-				<dependency id="linq2db"                                   version="3.6.0"     />
+				<dependency id="LINQPad.Reference"                         version="1.3.0"     />
+				<dependency id="linq2db"                                   version="4.0.0"     />
```
[the remark this answers is not in the document]
Comment (MaceWindu) [reply Build/linq2db.LINQPad.nuspec:15]: still not updated :smile:

[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Build/linq2db.LINQPad.nuspec]: 😂🤦🏼‍♂️

File: Build/linq2db.LINQPad.nuspec [hunk 8]
```diff
@@ -11,26 +11,26 @@
...
-				<dependency id="Humanizer.Core"                            version="2.13.14"   />
-				<dependency id="CodeJam"                                   version="4.0.1"     />
+				<dependency id="Humanizer.Core"                            version="2.14.1"    />
+				<dependency id="CodeJam"                                   version="4.1.0"     />
 				<dependency id="System.Configuration.ConfigurationManager" version="6.0.0"     />
-				<dependency id="Microsoft.CodeAnalysis.CSharp"             version="4.0.1"     />
-				<dependency id="System.Text.Json"                          version="5.0.2"     />
+				<dependency id="Microsoft.CodeAnalysis.CSharp"             version="4.2.0"     />
+				<dependency id="System.Text.Json"                          version="6.0.4"     />
 				<dependency id="dotMorten.Microsoft.SqlServer.Types"       version="1.3.0"     />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:22]: 1.5.0 is available

File: Build/linq2db.LINQPad.nuspec [hunk 9]
```diff
@@ -11,26 +11,26 @@
...
-				<dependency id="CodeJam"                                   version="4.0.1"     />
+				<dependency id="Humanizer.Core"                            version="2.14.1"    />
+				<dependency id="CodeJam"                                   version="4.1.0"     />
 				<dependency id="System.Configuration.ConfigurationManager" version="6.0.0"     />
-				<dependency id="Microsoft.CodeAnalysis.CSharp"             version="4.0.1"     />
-				<dependency id="System.Text.Json"                          version="5.0.2"     />
-				<dependency id="dotMorten.Microsoft.SqlServer.Types"       version="1.3.0"     />
+				<dependency id="Microsoft.CodeAnalysis.CSharp"             version="4.2.0"     />
+				<dependency id="System.Text.Json"                          version="6.0.4"     />
+				<dependency id="dotMorten.Microsoft.SqlServer.Types"       version="1.5.0"     />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:22]: these 3 deps missing from release notes

File: Build/linq2db.LINQPad.nuspec [hunk 10]
```diff
@@ -11,26 +11,26 @@
...
 
-				<dependency id="FirebirdSql.Data.FirebirdClient"           version="8.5.4"     />
-				<dependency id="MySqlConnector"                            version="2.0.0"     />
+				<dependency id="FirebirdSql.Data.FirebirdClient"           version="9.0.1"     />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:24]: missing from release notes

File: Build/linq2db.LINQPad.nuspec [hunk 11]
```diff
@@ -11,26 +11,26 @@
...
+				<dependency id="MySqlConnector"                            version="2.1.10"    />
 				<dependency id="AdoNetCore.AseClient"                      version="0.19.2"    />
-				<dependency id="System.Data.SQLite.Core"                   version="1.0.115.5" />
+				<dependency id="System.Data.SQLite.Core"                   version="1.0.116"   />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:27]: missing from release notes

File: Build/linq2db.LINQPad.nuspec [hunk 12]
```diff
@@ -11,26 +11,26 @@
...
 				<dependency id="System.Data.SqlClient"                     version="4.8.3"     />
 				<dependency id="System.Data.Odbc"                          version="5.0.0"     />
 				<dependency id="System.Data.OleDb"                         version="5.0.0"     />
-				<dependency id="Npgsql"                                    version="6.0.0"     />
-				<dependency id="Oracle.ManagedDataAccess.Core"             version="3.21.4"    />
-				<dependency id="IBM.Data.DB2.Core"                         version="3.1.0.500" />
+				<dependency id="Npgsql"                                    version="6.0.4"     />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:31]: missing from release notes

File: Build/linq2db.LINQPad.nuspec [hunk 13]
```diff
@@ -11,26 +11,26 @@
...
+				<dependency id="Oracle.ManagedDataAccess.Core"             version="3.21.61"   />
+				<dependency id="IBM.Data.DB2.Core"                         version="3.1.0.600" />
```
Comment (MaceWindu) [inline Build/linq2db.LINQPad.nuspec:33]: missing from release notes

File: Directory.Packages.props [hunk 14]
```diff
@@ -1,42 +1,43 @@
...
 	</ItemGroup>
 	
 	<ItemGroup Condition=" '$(TargetFramework)' == 'net461'">
-		<PackageVersion Include="Microsoft.CodeAnalysis.CSharp"   Version="2.10.0" />
-		<PackageVersion Include="Npgsql"                          Version="4.1.10" />
-		<PackageVersion Include="FirebirdSql.Data.FirebirdClient" Version="7.10.1" />
+		<PackageVersion Include="Microsoft.CodeAnalysis.CSharp"   Version="2.10.0"  />
+		<PackageVersion Include="Npgsql"                          Version="4.1.10"  />
+		<PackageVersion Include="FirebirdSql.Data.FirebirdClient" Version="7.10.1"  />
+		<PackageVersion Include="Oracle.ManagedDataAccess"        Version="19.13.0" />
```
Comment (MaceWindu) [inline Directory.Packages.props:33]: 1. no need to move it to conditioned item group
2. this file doesn't work well with `dotnet build` if package is not in unconditioned group

File: release-notes.md [hunk 15]
```diff
@@ -1,3 +1,14 @@
+# Release 4.0.0
+- [#55](https://github.com/linq2db/linq2db.LINQPad/pull/65) Fix bug in pluralization/singularization code
+- [ALL] dependency update: linq2db 3.6.0 -> 4.0.1
+- [ALL] dependency update: LINQPAD.Reference 1.1.0 -> 1.3.0
+- [ALL] dependency update: Humanizer.Core 2.13.14 -> 2.14.1
+- [ALL] dependency update: CodeJam 4.0.1 -> 4.1.0
+- [ALL] dependency update: System.Memory 4.5.4 -> 4.5.5
+- [ALL] dependency update: MySqlConnector 2.0.0 -> 2.1.10
+- [ALL] dependency update: Oracle.ManagedDataAccess.Core 3.21.4 -> 3.21.61
+- [LINQPAD7] dependency update: Oracle.ManagedDataAccess 19.13.0 -> 21.6.1
```
Comment (MaceWindu) [inline release-notes.md:10]: add reference to https://github.com/linq2db/linq2db.LINQPad/issues/63

**Review by MaceWindu — CHANGES_REQUESTED** [review]
image

File: release-notes.md (hunk 15 above)
Comment (MaceWindu) [reply release-notes.md:10]: Ah, also check affected LINQPAD version for dependencies. I don't think all of them updated for `ALL`

File: Directory.Packages.props (hunk 14 above)
Comment (viceroypenguin) [reply Directory.Packages.props:33]: 21.6.1 is locked to 4.6.2 or higher. was not aware it was netfx only, so I moved it. choice is to update to 4.6.2/4.8.1 or to use older version of oracle.manageddataaccess

Comment (MaceWindu) [reply Directory.Packages.props:33]: I see, let's stick to 19.14.0 then

Comment (MaceWindu) [reply Directory.Packages.props:33]: we release only one netfx build so it doesn't make sense to reference both versions

File: release-notes.md [hunk 16]
```diff
@@ -1,3 +1,21 @@
...
+- [#63](https://github.com/linq2db/linq2db.LINQPad/pull/63) Update Oracle.ManagedDataAccess.Core dependency to fix warning (thanks @db2222)
+- [#65](https://github.com/linq2db/linq2db.LINQPad/pull/65) Fix bug in pluralization/singularization code
+- [ALL] dependency update: linq2db 3.6.0 -> 4.0.1
+- [ALL] dependency update: LINQPAD.Reference 1.1.0 -> 1.3.0
+- [ALL] dependency update: Humanizer.Core 2.13.14 -> 2.14.1
+- [ALL] dependency update: CodeJam 4.0.1 -> 4.1.0
+- [ALL] dependency update: System.Memory 4.5.4 -> 4.5.5
+- [ALL] dependency update: MySqlConnector 2.0.0 -> 2.1.10
+- [ALL] dependency update: System.Data.SQLite.Core 1.0.115.5 -> 1.0.116
+- [ALL] dependency update: Oracle.ManagedDataAccess.Core 3.21.4 -> 3.21.61
```
Comment (MaceWindu) [inline release-notes.md:11]: this is linqpad7 dependency

File: release-notes.md [hunk 17]
```diff
@@ -1,3 +1,21 @@
...
+- [ALL] dependency update: IBM.Data.DB2.Core 3.1.0.500 -> 3.1.0.600
```
Comment (MaceWindu) [inline release-notes.md:12]: this is linqpad7 dependency

File: release-notes.md [hunk 18]
```diff
@@ -1,3 +1,21 @@
...
+- [ALL] dependency update: dotMorten.Microsoft.SqlServer.Types 1.3.0 -> 1.5.0
```
Comment (MaceWindu) [inline release-notes.md:13]: this is linqpad 7 dependency

File: Directory.Packages.props [hunk 19]
```diff
@@ -1,42 +1,42 @@
...
 		<PackageVersion Include="System.Configuration.ConfigurationManager" Version="6.0.0"                                   />
-		<PackageVersion Include="MySqlConnector"                            Version="2.0.0"                                   />
+		<PackageVersion Include="MySqlConnector"                            Version="2.1.10"                                  />
 		<PackageVersion Include="AdoNetCore.AseClient"                      Version="0.19.2"                                  />
 		<PackageVersion Include="IBM.Data.DB.Provider"                      Version="11.5.5010.4" GeneratePathProperty="true" />
-		<PackageVersion Include="System.Data.SQLite.Core"                   Version="1.0.115.5"                               />
+		<PackageVersion Include="System.Data.SQLite.Core"                   Version="1.0.116"                                 />
 		<PackageVersion Include="System.Data.SqlClient"                     Version="4.8.3"                                   />
 		<PackageVersion Include="Oracle.ManagedDataAccess"                  Version="19.13.0"                                 />
```
Comment (MaceWindu) [inline Directory.Packages.props:20]: let's update  to 19.14.0

**Review by MaceWindu — CHANGES_REQUESTED** [review]
almost there :)

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**